internal/audit/cerboshub/cerboshub.go

```diff
@@ -0,0 +1,237 @@
+	DecisionSyncPrefix = []byte("bsdec")
+)
+
+func init() {
+	audit.RegisterBackend(Backend, func(_ context.Context, confW *config.Wrapper, decisionFilter audit.DecisionLogEntryFilter) (audit.Log, error) {
+		conf := new(Conf)
+		if err := confW.GetSection(conf); err != nil {
+			return nil, fmt.Errorf("failed to read cerboshub audit log configuration: %w", err)
+		}
+
+		syncer := NewIngestSyncer()
+
+		return NewLog(conf, decisionFilter, syncer)
+	})
+}
+
+type Log struct {
+	*local.Log
+	logger  *zap.Logger
+	syncer  IngestSyncer
+	ticker  *time.Ticker
+	trigger chan local.TriggerSignal
+}
+
+func NewLog(conf *Conf, decisionFilter audit.DecisionLogEntryFilter, syncer IngestSyncer) (*Log, error) {
+	log, err := local.NewLog(&conf.Conf, decisionFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := zap.L().Named("auditlog").With(zap.String("backend", Backend))
+
+	maxBatchSize := int(conf.Ingest.MaxBatchSize)
+	flushInterval := conf.Ingest.FlushInterval
+	flushTimeout := conf.Ingest.FlushTimeout
+	numGo := int(conf.Ingest.NumGoRoutines)
+
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	l := &Log{
+		Log:     log,
+		logger:  logger,
+		syncer:  syncer,
+		ticker:  ticker,
+		trigger: make(chan local.TriggerSignal, 1),
+	}
+
+	go local.StartTriggerLoop(ticker, l.trigger, l.StopChan)
+
+	l.Wg.Add(1)
+	go l.batchSyncer(maxBatchSize, numGo, flushInterval, flushTimeout)
+
+	return l, nil
+}
+
+func (l *Log) WriteAccessLogEntry(ctx context.Context, record audit.AccessLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	if err := l.Log.WriteAccessLogEntry(ctx, func() (*auditv1.AccessLogEntry, error) {
+		return rec, nil
+	}); err != nil {
+		return err
+	}
+
+	callID, err := audit.ID(rec.CallId).Repr()
+	if err != nil {
+		return fmt.Errorf("invalid call ID: %w", err)
+	}
+
+	key := local.GenKey(AccessSyncPrefix, callID)
+
+	return l.Write(ctx, key, nil)
+}
+
+func (l *Log) WriteDecisionLogEntry(ctx context.Context, record audit.DecisionLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	if err := l.Log.WriteDecisionLogEntry(ctx, func() (*auditv1.DecisionLogEntry, error) {
+		return rec, nil
+	}); err != nil {
+		return err
+	}
+
+	callID, err := audit.ID(rec.CallId).Repr()
+	if err != nil {
+		return fmt.Errorf("invalid call ID: %w", err)
+	}
+
+	key := local.GenKey(DecisionSyncPrefix, callID)
+
+	return l.Write(ctx, key, nil)
+}
+
+func (l *Log) batchSyncer(maxBatchSize, numGo int, flushInterval, flushTimeout time.Duration) {
+	logger := l.logger.With(zap.String("component", "syncer"))
+
+	for i := 0; i < goroutineResetThreshold; i++ {
+		select {
+		case <-l.StopChan:
+			l.Wg.Done()
+			return
+		case sig := <-l.trigger:
+			// BadgerDB transactions work with snapshot isolation so we only take a view of the DB.
+			// Subsequent writes aren't blocked.
+			stream := l.Db.NewStream()
+			stream.NumGo = numGo
+			stream.Prefix = SyncStatusPrefix
+
+			{
```

Comment:
Is it possible to convert these blocks to functions? It's quite easy to "forget" you're inside a block (which I did several times 😄 ) and that could be a source of  bugs in the future.